flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/api/java/internal/StreamTableEnvironmentImpl.java

```diff
@@ -305,6 +327,11 @@ protected void validateTableSource(TableSource<?> tableSource) {
 		validateTimeCharacteristic(TableSourceValidation.hasRowtimeAttribute(tableSource));
 	}
 
+	@Override
+	protected boolean shouldTranslateEagerly() {
+		return true;
```

Comment:
blink planner will also use this class, if i understand correctly. so here could not return true deterministically.

Reply:
Correct, we wanted to leave it this way. The thing is if we use this class we should not allow lazy translation. All `toDataStream` conversion must be translated eagerly. If we allowed for other transformations to be executed lazily then some parts would be executed lazily and others eagerly. This poses many problems. How do we handle DML? DCLs commands if some are executed eagerly and others lazily.

Another problem is what happens if the user calls `execute` on `StreamExecutionEnvironment`, which queries will be executed?

Reply:
make sense

Reply:
`<T, ACC> void registerFunction(String name, AggregateFunction<T, ACC> aggregateFunction);` and `<T> void registerFunction(String name, TableFunction<T> tableFunction);` are only defined in `StreamTableEnvironment`, not in `TableEnvironment`. Should they be moved to  `TableEnvironment` to support lazy translation?

Reply:
We cannot do that. Those interface depend on type extraction (java/scala specific). We cannot add them to the unified interface. This has to wait until we have the new type inference for udfs in place.

Reply:
maybe the plan result of tests which use `registerFunction(String name, AggregateFunction<T, ACC> aggregateFunction)` and ` registerFunction(String name, TableFunction<T> tableFunction)` will be updated in blink planner,  and be reverted back when new type inference is ready.